### PR TITLE
feat:add op-geth feature

### LINKED ISSUE #6: Make L2 op-geth bare metal deployment on local machine
This is could be a sub task for #3

### PR DESCRIPTION
**Pull request by GrapeBaBa**
fix #6

### CODE REVIEW

**Comment by GrapeBaBa** [comment]
maybe some duplicated code(l1 geth and l2 geth is similar code) could be refactor later.

File: roll.py [hunk 1]
```diff
@@ -30,6 +30,10 @@
     "l1",
     help="spins up a local L1 node with the rollup contracts deployed on it")
 
+subparsers.add_parser(
+    "l2",
+    help="spins up a local op-geth node")
```
Comment (norswap) [inline roll.py:35]: I know this is the naming scheme for the Optimism repo but I suggest we switch to something more descriptive like "l2-execution". Then we can use "l2" as the target to deploy every part of an L2 (execution, proposer, batcher).

File: setup.py [hunk 2]
```diff
@@ -48,4 +51,32 @@ def setup_optimism_repo():
...
 
+
 ####################################################################################################
+
+def setup_op_geth_repo():
+    """
+    Clone the op-geth repository and build it.
+    """
+    github_url = "https://github.com/ethereum-optimism/op-geth.git"
+    git_tag = "v1.101106.0"
```
Comment (norswap) [inline setup.py:62]: We should an issue to add functionality to update the optimism and op-geth repo to new version.

File: setup.py [hunk 3]
```diff
@@ -48,4 +51,32 @@ def setup_optimism_repo():
...
+        raise Exception("Error: 'op-geth' exists as a file and not a directory.")
+    elif not os.path.exists("op-geth"):
+        descr = "clone the op-geth repository"
+        lib.run(descr, f"git clone {github_url}")
+        print(f"Succeeded: {descr}")
+
+    lib.run("checkout stable version", f"git checkout --detach {git_tag}",
+            cwd="op-geth")
+
+    print("Starting to build the op-geth repository. This may take a while...\n")
```
Comment (norswap) [inline setup.py:74]: I like to add the location of the log file to these messages.

File: l1.py [hunk 4]
```diff
@@ -398,9 +346,12 @@ def clean(paths: OPPaths):
     Cleans up build outputs, such that trying to deploy the devnet L1 node will proceed as though it
     was the first invocation.
     """
-    if os.path.exists(paths.devnet_l1_gen_dir):
-        print(f"Cleaning up {paths.devnet_l1_gen_dir}")
-        shutil.rmtree(paths.devnet_l1_gen_dir, ignore_errors=True)
+    if os.path.exists(paths.devnet_gen_dir):
+        print(f"Cleaning up {DEVNET_L1_DATA_DIR}")
```
Comment (norswap) [inline l1.py:350]: Should be `{paths.devnet_gen_dir}` in that string.

Do you think this need to be printed every time we run clean?

File: l2.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+import os
+import shutil
+import subprocess
+import sys
+
+import libroll as lib
+from paths import OPPaths
+from processes import PROCESS_MGR
+
+
```
Comment (norswap) [inline l2.py:14]: I would drop the "devnet"  part of the names everywhere that is not strictly "devnet" (i.e. a local developer deployment) related. So any logic we can reuse for production deployment (which ideally should be most of it) shouldn't say "devnet".

File: l2.py [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+####################################################################################################
+
+
+def clean(paths: OPPaths):
+    """
+    Cleans up build outputs, such that trying to deploy the devnet op-geth node will proceed as though it
+    was the first invocation.
+    """
+    if os.path.exists(paths.devnet_gen_dir):
+        print(f"Cleaning up {paths.devnet_gen_dir}")
```
Comment (norswap) [inline l2.py:227]: Same comment as in other file, maybe make debug statement?

File: l2.py [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+            f"--ws.port={cfg.ws_port}",
+            "--ws.origins=*",
+            "--ws.api=debug,eth,txpool,net,engine",
+
+            # Configuration for clique signing, clique itself is enabled via the genesis file
+            f"--unlock={cfg.signer_address}",
+            "--mine",
+            f"--miner.etherbase={cfg.signer_address}",
+            f"--password={cfg.data_dir}/password",
+            "--allow-insecure-unlock",
```
Comment (norswap) [inline l2.py:197]: This whole block of options is not needed on L2.

File: l2.py [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+        """Port to use for the WebSocket-based JSON_RPC server."""
+
+
+####################################################################################################
+
+def start_devnet_l2_node(paths: OPPaths):
+    """
+    Spin the devnet op-geth node, then wait for it to be ready.
+    """
+
```
Comment (norswap) [inline l2.py:112]: The version of the OP repo we were using used the same entrypoint shell script for L1 and L2, but quite a few of the options are not actually needed on L2. The lastest version of the Optimism repo has them split out:

- https://github.com/ethereum-optimism/optimism/blob/9c1ee08c0606d067a9754c4b6d798933fbe01d3a/ops-bedrock/entrypoint-l1.sh
- https://github.com/ethereum-optimism/optimism/blob/9c1ee08c0606d067a9754c4b6d798933fbe01d3a/ops-bedrock/entrypoint-l2.sh

Will comment as appropriate!

File: l2.py [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+            f.write(cfg.password)
+        with open(cfg.tmp_signer_key_path, "w") as f:
+            f.write(cfg.signer_private_key.replace("0x", ""))
+        lib.run(
+            "importing signing keys",
+            ["op-geth", "account", "import",
+             f"--datadir={cfg.data_dir}",
+             f"--password={cfg.password_path}",
+             cfg.tmp_signer_key_path])
+        os.remove(f"{cfg.data_dir}/block-signer-key")
```
Comment (norswap) [inline l2.py:141]: Apparently the whole password thing is not needed for the L2 side? That would mean the signing is done by the op-node.

Could you test if that is indeed the case? (If not, it means it was changed after the old commit we're using.)

File: l2.py [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+            f"--datadir={cfg.data_dir}",
+            f"--verbosity={cfg.verbosity}",
+
+            f"--networkid={cfg.chain_id}",
+            "--syncmode=full",  # doesn't matter, it's only us
+            "--gcmode=archive",
+
+            # No peers: the blockchain is only this node
+            "--nodiscover",
+            "--maxpeers=1",
```
Comment (norswap) [inline l2.py:175]: This can be 0 on L2.

File: l2.py (hunk 8 above)
Comment (norswap) [reply l2.py:112]: I also see that upnode has added the following options:

```
  --rollup.disabletxpoolgossip=true \
  --txlookuplimit=0
```

(from https://github.com/upnodedev/op-stack-basic-deployment/blob/main/entrypoints/op-geth.sh)

We should investigate, see what these do, if they exist in the version we use.

File: l2.py [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+    def __init__(self, geth_data_dir: str, paths: OPPaths):
+        self.data_dir = geth_data_dir
+        """Geth data directory for devnet op-geth node."""
+
+        self.keystore_dir = f"{self.data_dir}/keystore"
+        """Keystore directory for devnet op-geth node (each file stores an encrypted signer key)."""
+
+        self.chaindata_dir = f"{self.data_dir}/geth/chaindata"
+        """Directory storing chain data."""
+
```
Comment (norswap) [inline l2.py:71]: Might want to remove some options here that become unused as a result of change in options highlighted in comments on the `start_devnet_l2_node` function.

File: setup.py (hunk 2 above)
Comment (norswap) [reply setup.py:62]: https://github.com/0xFableOrg/simple-op-stack-rollup/issues/29

File: l1.py (hunk 4 above)
Comment (GrapeBaBa) [reply l1.py:350]: it is ok for me for print every time

File: l2.py (hunk 6 above)
Comment (GrapeBaBa) [reply l2.py:227]: not get you here.

File: l1.py (hunk 4 above)
Comment (norswap) [reply l1.py:350]: The way I'm thinking about this is we should only print things that are useful for the people to know. They probably don't even need to know what these directories are. Ideally they just run the command, the things runs and it's done.

I'm not sure what they can possibly do that is useful when they see "Cleanup up `db/devnet/l1`"

File: l2.py (hunk 6 above)
Comment (norswap) [reply l2.py:227]: I mean same as https://github.com/0xFableOrg/simple-op-stack-rollup/pull/27#discussion_r1304163295 (remove prints or turn them into `lib.debug`

File: l1.py (hunk 4 above)
Comment (GrapeBaBa) [reply l1.py:350]: actually i am not sure make 'clean' in separate files if is good? should we just make clean logic `roll.py`

File: roll.py (hunk 1 above)
Comment (norswap) [reply roll.py:35]: I think l2-execution is probably a little bit more clear, but both are really fine by me.

Comment (norswap) [reply roll.py:35]: I'll take care of this later!

File: l1.py (hunk 4 above)
Comment (norswap) [reply l1.py]: Agree we should do that!

File: l2.py (hunk 8 above)
Comment (GrapeBaBa) [reply l2.py]: txlookuplimit default value is one year which i think it is enough

**Review by norswap — APPROVED** [review]

**Comment by norswap** [comment]
Fantastic work, let's goooo 🚀🚀🚀

**Pull request merged**